velox/jni/cpp/native/NativePlanBuilder.cc

```diff
@@ -0,0 +1,183 @@
+}
+
+void NativePlanBuilder::nativeProject(
+    JNIEnv* env,
+    jobject obj,
+    jobjectArray projections) {
+  JNI_METHOD_START
+  const std::shared_ptr<PlanBuilder> builder = as<PlanBuilder>(obj);
+  builder->project(ConvertJStringArrayToVector(env, projections));
+  JNI_METHOD_END()
+}
+
+jstring
+NativePlanBuilder::nativeJavaScan(JNIEnv* env, jobject obj, jstring schema) {
+  JNI_METHOD_START
+  const std::shared_ptr<PlanBuilder> builder = as<PlanBuilder>(obj);
+  const std::string schemaStr = jStringToCString(env, schema);
+  const RowTypePtr rowType = asRowType(ExprUtils::toVeloxType(schemaStr));
+  builder->tableScan(rowType);
+  core::PlanNodeId scanNodeId;
+  builder->capturePlanNodeId(scanNodeId);
+  return env->NewStringUTF(scanNodeId.c_str());
+  JNI_METHOD_END(nullptr)
+}
+
+void NativePlanBuilder::nativeFilter(JNIEnv* env, jobject obj, jstring filter) {
+  JNI_METHOD_START
+  const std::string filterStr = jStringToCString(env, filter);
+  const std::shared_ptr<PlanBuilder> builder = as<PlanBuilder>(obj);
+  auto previousHook = core::Expressions::getResolverHook();
+  parse::registerTypeResolver();
+  builder->filter(filterStr);
+  core::Expressions::setTypeResolverHook(previousHook);
+  JNI_METHOD_END()
+}
+jstring NativePlanBuilder::nativeWindowFunction(
+    JNIEnv* env,
+    jobject obj,
+    jstring functionCall,
+    jstring fram,
+    jboolean ignoreNullKey) {
+  JNI_METHOD_START
+  auto typePtr = ISerializable::deserialize<core::ITypedExpr>(
+      folly::parseJson(
+          jStringToCString(env, functionCall), getSerializationOptions()),
+      sdk::memory::MemoryManager::get()->planMemoryPool().get());
+  std::shared_ptr<const core::CallTypedExpr> callExpr =
+      std::dynamic_pointer_cast<const core::CallTypedExpr>(typePtr);
+  auto frame = core::WindowNode::Frame::deserialize(
+      folly::parseJson(jStringToCString(env, fram), getSerializationOptions()));
+  bool ignoreNull = ignoreNullKey;
+  return env->NewStringUTF(
+      folly::toJson(
+          core::WindowNode::Function{callExpr, frame, ignoreNull}.serialize())
+          .c_str());
+  JNI_METHOD_END(env->NewStringUTF(""))
+}
+
+void NativePlanBuilder::nativeLimit(
+    JNIEnv* env,
+    jobject obj,
+    jint offset,
+    jint limit) {
+  JNI_METHOD_START
+  const std::shared_ptr<PlanBuilder> builder = as<PlanBuilder>(obj);
+  builder->limit(offset, limit, false);
+  JNI_METHOD_END()
+}
+
+jstring NativePlanBuilder::nativeNodeId(JNIEnv* env, jobject obj) {
+  JNI_METHOD_START
+  const std::shared_ptr<PlanBuilder> builder = as<PlanBuilder>(obj);
+  return env->NewStringUTF(builder->planNode()->id().c_str());
+  JNI_METHOD_END(nullptr)
+}
+
+jstring NativePlanBuilder::nativeBuilder(JNIEnv* env, jobject obj) {
+  JNI_METHOD_START
+  const std::shared_ptr<PlanBuilder> builder = as<PlanBuilder>(obj);
+  return env->NewStringUTF(
+      utils::JsonUtils::toSortedJson(builder->planNode()->serialize()).c_str());
+  JNI_METHOD_END(env->NewStringUTF("ERROR"))
```

Comment:
For such statements, using `JNI_METHOD_END(nullptr)` would be fine as exception is expected to throw to Java. Add meaningful items may happen to confuse developers.